python/ray/data/_internal/planner/plan_write_op.py

```diff
@@ -106,7 +107,19 @@ def _plan_write_op_internal(
 
     map_transformer = MapTransformer(transform_fns)
 
-    return MapOperator.create(
+    # Set up on_start callback for datasinks.
+    # This allows on_write_start to receive the schema from the first input bundle,
+    # enabling schema-dependent initialization (e.g., Iceberg schema evolution).
+    # NOTE: _FileDatasink is excluded because dataset.py already calls on_write_start()
+    # explicitly before execution to handle SaveMode checks and directory creation.
+    on_start = None
+    if isinstance(datasink, Datasink):
+        from ray.data.datasource.file_datasink import _FileDatasink
+
+        if not isinstance(datasink, _FileDatasink):
+            on_start = datasink.on_write_start
```

Comment:
See comment below -- let's not special-case it, consolidate it to just a single flow